Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,212 @@
+A lot of care needs to be taken when translating between local and global indices since any mistake will lead to incorrect results that can be very hard notice and are very difficult to trace down.
+It is thus absolutely crucial that this is done only in a single location in our code base and is unit tested well.
+
+#### Instrument 2.0:
+
+Instrument 2.0 currently under [prototype](https://github.com/DMSC-Instrument-Data/instrument-prototype) will come with MPI support as per the [design](https://github.com/DMSC-Instrument-Data/documents/tree/master/meeting_notes/April_2016/InstrumentPrototype#mpi-support).
+Cleaning up our index handling is thus a prerequisite for its implementation.
+
+
+## Detailed goals
+
+- All algorithms should have a consistent interface.
+- Eliminate duplicated code for declaring properties for index ranges and list.
+- Eliminate duplicated code for validating indices and converting them to the desired type.
+- Developers should generally not have to deal with spectrum number or detector ID at all.
+  Spectrum number and detector ID are important for the user interface (or file interface), but irrelevant for the internals of most algorithms.
+- Developers must be isolated from dealing explicitly with the local and global indices that come with an MPI implementation of Mantid.
+
+## User interface
+
+On the user-interface side it must be possible to specify certain spectra or detectors.
+
+#### SpectrumNumbersProperty
+
+- The interface of algorithms should make it possible to specify spectrum number ranges and lists (full range, min/max, list).
+- For better compatibility for the current interface and for users who do not like to deal with spectrum numbers it should also be possible to specify workspace indices.
+  It is important to note that this is a **global** workspace index.
+  As explained above, in an MPI-based Mantid run this is **not** the same as the workspace index.
+- Some algorithms also accept detector IDs and then identify corresponding spectrum numbers.
+
+For developers there should be single property that provides all this:
+
+- When declaring the property it must be possible to specify which of the index types the algorithm should accept.
+- A reasonable default would be *spectrum number* and *global workspace index*.
+- *Detector ID* is used in few cases and should be optional.
+- The word *global* should most likely be omitted from the user interface.
+  The translation from global to local workspace indices happens between the UI and the algorithm, i.e., neither the user nor (generally) the developer needs to know about global indices.
+  Putting *global* in the UI would only create confusion.
+
+#### DetectorIDsProperty
+
+Some algorithms work directly with detectors and thus need a detector index.
+This is similar to the `SpectrumNumbersProperty` but simpler.
+Most likely it would be sufficient to provide only *detector ID* as possible input.
+
+It should be pointed out that:
+
```

Comment:
One breaking change would be naming of properties, if we want to enforce consistency.

Regarding ambiguous interpretation, there is [17073](https://github.com/mantidproject/mantid/issues/17073) (if I understand the problem correctly).